File: server/lib/singbox.mjs
import fs from 'node:fs';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import crypto from 'node:crypto';

export function keyForMethod(method) { if (method.includes('aes-128')) return crypto.randomBytes(16).toString('base64'); return crypto.randomBytes(32).toString('base64'); }
export function credentialsForClient(client) { return client.credentials || { reality: crypto.randomUUID(), hysteria2: crypto.randomBytes(18).toString('base64url'), shadowsocks2022: crypto.randomBytes(18).toString('base64url') }; }
function clientsFor(item, clients) { return clients.filter(x => x.enabled !== false && Array.isArray(x.inbounds) && x.inbounds.includes(item.id)); }
export function inboundToSingBox(item, clients = [], settings = {}) {
  const selected = clientsFor(item, clients);
  const base = { tag: item.name, listen: item.listen || '0.0.0.0', listen_port: Number(item.port) };
  if (item.type === 'reality') return { type: 'vless', ...base, users: selected.map(client => ({ name: client.username, uuid: credentialsForClient(client).reality })), tls: { enabled: true, server_name: item.handshakeServer, reality: { enabled: true, handshake: { server: item.handshakeServer, server_port: Number(item.handshakePort || 443) }, private_key: item.privateKey, short_id: [item.shortId] } } };
  if (item.type === 'hysteria2') return { type: 'hysteria2', ...base, up_mbps: Number(item.upMbps || 0), down_mbps: Number(item.downMbps || 0), users: selected.map(client => ({ name: client.username, password: credentialsForClient(client).hysteria2 })), obfs: item.obfsPassword ? { type: 'salamander', password: item.obfsPassword } : undefined, tls: { enabled: true, server_name: item.tlsServerName, certificate_path: item.certificatePath || settings.certFullchainPath || '/var/lib/kotaui/certs/fullchain.pem', key_path: item.keyPath || settings.certPrivateKeyPath || '/var/lib/kotaui/certs/privkey.pem' }, masquerade: item.masquerade || undefined };
  if (item.type === 'shadowsocks2022') return { type: 'shadowsocks', ...base, network: item.network || undefined, method: item.method, password: item.password || keyForMethod(item.method), users: selected.map(client => ({ name: client.username, password: credentialsForClient(client).shadowsocks2022 })) };
  throw new Error(`unsupported inbound type: ${item.type}`);
}
export function buildConfig(state, apiPort = 9090) { return { '$schema': 'https://sing-box.sagernet.org/schema.json', log: { level: 'info' }, inbounds: state.inbounds.filter(x => x.enabled).map(x => inboundToSingBox(x, state.clients, state.settings)), outbounds: [{ type: 'direct', tag: 'direct' }], experimental: { v2ray_api: { listen: `127.0.0.1:${apiPort}`, stats: { enabled: true, inbounds: state.inbounds.filter(x => x.enabled).map(x => x.name), users: state.clients.map(x => x.username) } } } }; }
export function writeConfig(state, configFile, apiPort = 9090) { const config = buildConfig(state, apiPort); fs.mkdirSync(path.dirname(configFile), { recursive: true, mode: 0o700 }); const tmp = `${configFile}.tmp`; fs.writeFileSync(tmp, JSON.stringify(config, null, 2), { mode: 0o600 }); fs.renameSync(tmp, configFile); return config; }
export function validateConfig(configFile, binary = 'sing-box') { const result = spawnSync(binary, ['check', '-c', configFile], { encoding: 'utf8' }); if (result.error?.code === 'ENOENT') return { ok: false, available: false, error: 'sing-box 未安装，无法执行真实配置校验' }; return { ok: result.status === 0, available: true, output: `${result.stdout || ''}${result.stderr || ''}`.trim() }; }

File: internal/proxy/singbox.go
package proxy

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net"
	"net/url"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/ptfpwcpzy/KotaUI/internal/config"
)

func Write(state config.State, runtime config.Runtime) error {
	if err := os.MkdirAll(filepath.Dir(runtime.SingBoxConfig), 0700); err != nil {
		return err
	}
	inbounds := make([]map[string]any, 0, len(state.Inbounds))
	active := activeClients(state.Clients)
	for _, inbound := range state.Inbounds {
		if !inbound.Enabled {
			continue
		}
		switch inbound.Type {
		case "reality":
			users := []map[string]any{}
			for _, client := range active {
				if secret := client.Credentials[inbound.ID]; secret != "" {
					users = append(users, map[string]any{"name": client.Username, "uuid": secret, "flow": "xtls-rprx-vision"})
				}
			}
			inbounds = append(inbounds, map[string]any{"type": "vless", "tag": inbound.ID, "listen": inbound.Listen, "listen_port": inbound.Port, "users": users, "tls": map[string]any{"enabled": true, "server_name": inbound.SNI, "reality": map[string]any{"enabled": true, "handshake": map[string]any{"server": inbound.HandshakeServer, "server_port": inbound.HandshakePort}, "private_key": inbound.PrivateKey, "short_id": []string{inbound.ShortID}}}})
		case "hysteria2":
			users := []map[string]any{}
			for _, client := range active {
				if secret := client.Credentials[inbound.ID]; secret != "" {
					users = append(users, map[string]any{"name": client.Username, "password": secret})
				}
			}
			tls := map[string]any{"enabled": true, "certificate_path": runtime.TLSCert, "key_path": runtime.TLSKey, "server_name": runtime.Domain}

			entry := map[string]any{"type": "hysteria2", "tag": inbound.ID, "listen": inbound.Listen, "listen_port": inbound.Port, "users": users, "tls": tls, "up_mbps": inbound.UpMbps, "down_mbps": inbound.DownMbps}
			if inbound.ObfsPassword != "" {
				entry["obfs"] = map[string]any{"type": "salamander", "password": inbound.ObfsPassword}
			}
			inbounds = append(inbounds, entry)
		case "shadowsocks2022":
			users := []map[string]any{}
			for _, client := range active {
				if secret := client.Credentials[inbound.ID]; secret != "" {
					users = append(users, map[string]any{"name": client.Username, "password": secret})
				}
			}
			inbounds = append(inbounds, map[string]any{"type": "shadowsocks", "tag": inbound.ID, "listen": inbound.Listen, "listen_port": inbound.Port, "method": "2022-blake3-aes-256-gcm", "password": inbound.ServerPassword, "users": users})
		}
	}
	statsUsers := make([]string, 0, len(state.Clients))
	for _, client := range state.Clients {
		if client.Username != "" {
			statsUsers = append(statsUsers, client.Username)
		}
	}
	sort.Strings(statsUsers)
	directOutbound := map[string]any{"type": "direct", "tag": "direct"}
	if strategy := state.Settings.OutboundStrategy; strategy != "" && strategy != "auto" {
		directOutbound["domain_strategy"] = strategy
	}
	root := map[string]any{
		"log":       map[string]any{"level": "warn", "timestamp": true},
		"inbounds":  inbounds,
		"outbounds": []map[string]any{directOutbound},
		"experimental": map[string]any{
			"v2ray_api": map[string]any{
				"listen": fmt.Sprintf("127.0.0.1:%d", runtime.StatsPort),
				"stats":  map[string]any{"enabled": true, "users": statsUsers},
			},
		},
	}
	body, err := json.MarshalIndent(root, "", "  ")
	if err != nil {
		return err
	}
	tmp := runtime.SingBoxConfig + ".tmp"
	if err := os.WriteFile(tmp, body, 0600); err != nil {
		return err
	}
	return os.Rename(tmp, runtime.SingBoxConfig)
}

func activeClients(clients []config.Client) []config.Client {
	now := time.Now()
	currentMonth := now.Format("2006-01")
	out := []config.Client{}
	for _, client := range clients {
		if client.Month != currentMonth {
			client.MonthlyUsedBytes = 0
		}
		if client.Active(now) {
			out = append(out, client)
		}
	}
	return out
}

func Subscription(state config.State, runtime config.Runtime, username string) (string, bool) {
	var client *config.Client
	for i := range state.Clients {
		if state.Clients[i].Username == username {
			client = &state.Clients[i]
			break
		}
	}
	if client == nil || !client.Active(time.Now()) {
		return "", false
	}
	links := []string{}
	for _, id := range client.InboundIDs {
		for _, inbound := range state.Inbounds {
			if inbound.ID != id || !inbound.Enabled {
				continue
			}
			secret := client.Credentials[id]
			if secret == "" {
				continue
			}
			label := urlLabel(inbound.Name, client.Username)
			switch inbound.Type {
			case "reality":
				host := shareHost(inbound, runtime)
				query := url.Values{}

				query.Set("encryption", "none")
				query.Set("security", "reality")
				query.Set("flow", "xtls-rprx-vision")
				query.Set("sni", inbound.SNI)
				query.Set("fp", "chrome")
				query.Set("pbk", inbound.PublicKey)
				query.Set("sid", inbound.ShortID)
				query.Set("type", "tcp")
				query.Set("headerType", "none")
				links = append(links, fmt.Sprintf("vless://%s@%s:%d?%s#%s", url.PathEscape(secret), host, inbound.Port, query.Encode(), url.QueryEscape(label)))
			case "hysteria2":
				query := url.Values{"sni": []string{runtime.Domain}, "insecure": []string{"0"}}
				if inbound.ObfsPassword != "" {
					query.Set("obfs", "salamander")
					query.Set("obfs-password", inbound.ObfsPassword)
				} else {
					query.Set("obfs", "none")
				}
				links = append(links, fmt.Sprintf("hysteria2://%s@%s:%d/?%s#%s", url.PathEscape(secret), shareHost(inbound, runtime), inbound.Port, query.Encode(), url.QueryEscape(label)))
			case "shadowsocks2022":
				// SS2022 multi-user mode requires the server EIH key followed by the user key.
				encoded := base64.RawStdEncoding.EncodeToString([]byte("2022-blake3-aes-256-gcm:" + inbound.ServerPassword + ":" + secret))
				links = append(links, fmt.Sprintf("ss://%s@%s:%d#%s", encoded, shareHost(inbound, runtime), inbound.Port, url.QueryEscape(label)))

			}
		}
	}
	return strings.Join(links, "\n"), true
}

func shareHost(inbound config.Inbound, runtime config.Runtime) string {
	if !inbound.UseIPv6 {
		return runtime.Domain
	}
	if ip := publicIPv6(); ip != "" {
		return "[" + ip + "]"
	}
	return runtime.Domain
}

func publicIPv6() string {
	interfaces, err := net.Interfaces()
	if err != nil {
		return ""
	}
	for _, iface := range interfaces {
		addresses, err := iface.Addrs()
		if err != nil {
			continue
		}
		for _, address := range addresses {
			var ip net.IP
			switch value := address.(type) {
			case *net.IPNet:
				ip = value.IP
			case *net.IPAddr:
				ip = value.IP
			}
			if ip != nil && ip.To4() == nil && ip.IsGlobalUnicast() && !ip.IsPrivate() && !ip.IsLinkLocalUnicast() {
				return ip.String()
			}
		}
	}
	return ""
}

func urlLabel(parts ...string) string {
	return strings.NewReplacer(" ", "-", "#", "", "?", "").Replace(strings.Join(parts, "-"))
}
